azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -0,0 +1,132 @@
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This provides the ability to cleanup up execution containers which exceed the maximum allowed
+ * duration for a flow. Following extensions will be considered in future. (1) Convert this to a
+ * general background process for any other containerization related periodic tasks. (2) Add ability
+ * to track containers which may not be have been deleted despite prior attempts. (3) It's assumed
+ * final status updates to the db are handled by other parts, such as flow-executor. That can also
+ * be added here as a failsafe if needed.
+ */
+@Singleton
+public class ContainerCleanupManager {
+
+  private static final Logger logger = LoggerFactory.getLogger(ContainerCleanupManager.class);
+  private static final Duration DEFAULT_STALE_CONTAINER_CLEANUP_INTERVAL = Duration.ofMinutes(10);
+  private static final Duration DEFAULT_STALE_CONTAINER_AGE = Duration.ofDays(10);
+  private final long cleanupIntervalMin;
+  private final long staleContainerAgeDays;
+  private final ScheduledExecutorService cleanupService;
+  private final ExecutorLoader executorLoader;
+  private final ContainerizedImpl containerizedImpl;
+
+  @Inject
+  public ContainerCleanupManager(final Props azkProps, final ExecutorLoader executorLoader,
+      final ContainerizedImpl containerizedImpl) {
+    this.cleanupIntervalMin = azkProps
+        .getLong(
+            ContainerizedExecutionManagerProperties.CONTAINERIZED_STALE_EXECUTION_CLEANUP_INTERVAL_MIN,
+            DEFAULT_STALE_CONTAINER_CLEANUP_INTERVAL.toMinutes());
+    this.staleContainerAgeDays = azkProps
+        .getLong(
+            ContainerizedExecutionManagerProperties.CONTAINERIZED_STALE_EXECUTION_AGE_DAYS,
+            DEFAULT_STALE_CONTAINER_AGE.toDays());
+    this.cleanupService = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder().setNameFormat("azk-container-cleanup").build());
+    this.executorLoader = executorLoader;
+    this.containerizedImpl = containerizedImpl;
+  }
+
+  /**
+   * Execute container-provider specific APIs for all 'stale' containers. A container is considered
+   * 'stale' if it was launched CONTAINERIZED_STALE_EXECUTION_AGE_DAYS ago and the corresponding
+   * execution is not yet in a final state.
+   * <p>
+   * It's important that this method does not throw exceptions as that will interrupt the scheduling
+   * of {@code cleanupService}.
+   */
+  public void terminateStaleContainers() {
+    try {
+      final List<ExecutableFlow> staleFlows = this.executorLoader.fetchStaleFlows(Duration.ofDays(
+          this.staleContainerAgeDays));
+      for (final ExecutableFlow flow : staleFlows) {
+        deleteContainerQuietly(flow.getExecutionId());
+      }
+    } catch (final Exception e) {
+      logger.error("Unexpected exception during container cleanup." + e);
```

Comment:
nit, no need to use concatenation. Exception could be the next param.

Reply:
Will do.